src/main/java/net/snowflake/client/jdbc/diagnostic/DiagnosticContext.java

```diff
@@ -0,0 +1,197 @@
+        String type = objectNode.get("type").asText();
+        String host = objectNode.get("host").asText();
+        int port = objectNode.get("port").asInt();
+        SnowflakeEndpoint e = new SnowflakeEndpoint(type, host, port);
+        endpoints.add(e);
+      }
+
+    } catch (IOException e) {
+      logger.error("Failed to read allowlist file: ", allowListFile);
+      logger.error(e.getLocalizedMessage(), e);
+    } catch (Exception e) {
+      logger.error("Failed to parse data in allowlist file:", allowListFile);
+      logger.error(e.getLocalizedMessage(), e);
+    }
+
+    tests =
+        new DiagnosticCheck[] {
+          new DnsDiagnosticCheck(proxyConf),
+          new TcpDiagnosticCheck(proxyConf),
+          new CertificateDiagnosticCheck(proxyConf),
+          new HttpAndHttpsDiagnosticCheck(proxyConf)
+        };
+  }
+
+  /** This constructor is only used for testing */
+  public DiagnosticContext(Map<SFSessionProperty, Object> connectionPropertiesMap) {
+    createProxyConfiguration(connectionPropertiesMap);
+
+    tests =
+        new DiagnosticCheck[] {
+          new DnsDiagnosticCheck(proxyConf),
+          new TcpDiagnosticCheck(proxyConf),
+          new CertificateDiagnosticCheck(proxyConf),
+          new HttpAndHttpsDiagnosticCheck(proxyConf)
+        };
+  }
+
+  private void createProxyConfiguration(Map<SFSessionProperty, Object> connectionPropertiesMap) {
+    String proxyHost = (String) connectionPropertiesMap.get(SFSessionProperty.PROXY_HOST);
+    int proxyPort =
+        (connectionPropertiesMap.get(SFSessionProperty.PROXY_PORT) == null)
+            ? -1
+            : Integer.parseInt((String) connectionPropertiesMap.get(SFSessionProperty.PROXY_PORT));
+    String nonProxyHosts = (String) connectionPropertiesMap.get(SFSessionProperty.NON_PROXY_HOSTS);
+    proxyConf = new ProxyConfig(proxyHost, proxyPort, nonProxyHosts);
+  }
+
+  public void runDiagnostics() {
+
+    getEnvironmentInfo();
+
+    // Loop through endpoints and run diagnostic test on each one of them
+    for (DiagnosticCheck test : tests) {
+      for (SnowflakeEndpoint endpoint : endpoints) {
+        test.run(endpoint);
+      }
+    }
+  }
+
+  private JsonNode readAllowListJsonFile(String jsonFilePath) throws IOException {
+    ObjectMapper objectMapper = new ObjectMapper();
+    File allowListFile = new File(jsonFilePath);
+
+    return objectMapper.readTree(allowListFile);
+  }
+
+  public void getEnvironmentInfo() {
+    logger.debug("Getting environment information");
+    logger.debug("Current truststore used: " + getTrustStoreLocation());
+    logger.debug("-Dnetworkaddress.cache.ttl: " + System.getProperty("networkaddress.cache.ttl"));
+    logger.debug(
+        "-Dnetworkaddress.cache.negative.ttl: "
+            + System.getProperty("networkaddress.cache.negative.ttl"));
+    logger.debug("-Djavax.net.debug: " + System.getProperty(JAVAX_NET_DEBUG));
+  }
+
+  private boolean isNullOrEmpty(String a) {
```

Comment:
I believe such function already exists in apache commons or some other library that we have ;)

Reply:
I was going for the approach of not using any dependencies in this package. Is it preferable to import `org.apache.commons.lang3.StringUtils` and use those utility methods? I guess either way if we ever want to get rid of Apache Commons for whatever reason then it's not going to be a big deal to refactor this.

Reply:
I think it's ok to have this method in our code and not use the external library - eventually we should minimize the number of dependencies